vector-stores/spring-ai-neo4j-store/src/main/java/org/springframework/ai/vectorstore/Neo4jVectorStore.java

```diff
@@ -0,0 +1,357 @@
+
+	private static final String INDEX_NAME = "spring-ai-document-index";
+
+	private static final String DEFAULT_EMBEDDING_PROPERTY = "embedding";
+
+	private final Driver driver;
+
+	private final EmbeddingClient embeddingClient;
+
+	private final Neo4jVectorStoreConfig config;
+
+	public Neo4jVectorStore(Driver driver, EmbeddingClient embeddingClient, Neo4jVectorStoreConfig config) {
+
+		Assert.notNull(driver, "Neo4j driver must not be null");
+		Assert.notNull(embeddingClient, "Embedding client must not be null");
+
+		this.driver = driver;
+		this.embeddingClient = embeddingClient;
+
+		this.config = config;
+	}
+
+	@Override
+	public void add(List<Document> documents) {
+
+		var rows = documents.stream().map(this::documentToRecord).toList();
+
+		try (var session = driver.session()) {
+			var statement = """
+						UNWIND $rows AS row
+						MERGE (u:%s {id: row.id})
+						ON CREATE
+							SET u += row.properties
+						ON MATCH
+							SET u = {}
+							SET u.id = row.id,
+								u += row.properties
+						WITH row, u
+						CALL db.create.setVectorProperty(u, $embeddingProperty, row.embedding)
+						YIELD node
+						RETURN count(node)
+					""".formatted(config.quotedLabel);
+			session.run(statement, Map.of("rows", rows, "embeddingProperty", config.embeddingProperty)).consume();
+		}
+	}
+
+	@Override
+	public Optional<Boolean> delete(List<String> idList) {
+
+		try (var session = driver.session(config.sessionConfig)) {
+
+			var summary = session.run("""
+					MATCH (n:%s) WHERE n.id IN $ids
+					CALL { WITH n DETACH DELETE n } IN TRANSACTIONS OF $transactionSize ROWS
+					 """.formatted(config.quotedLabel), Map.of("ids", idList, "transactionSize", 10_000)).consume();
+			return Optional.of(idList.size() == summary.counters().nodesDeleted());
+		}
+	}
+
+	@Override
+	public List<Document> similaritySearch(String query) {
+		return similaritySearch(query, 5);
+	}
+
+	@Override
+	public List<Document> similaritySearch(String query, int k) {
+		return similaritySearch(query, k, 0);
+	}
+
+	@Override
+	public List<Document> similaritySearch(String query, int k, double threshold) {
+
+		Assert.isTrue(k > 0, "The number of documents to returned must be greater than zero");
+		Assert.isTrue(threshold >= 0 && threshold <= 1,
+				"The similarity score is bounded between 0 and 1; least to most similar respectively.");
+
+		var embedding = Values.value(toFloatArray(this.embeddingClient.embed(query)));
+		try (var session = driver.session(config.sessionConfig)) {
+			return session
+				.run("""
+						CALL db.index.vector.queryNodes($indexName, $numberOfNearestNeighbours, $embeddingValue)
+						YIELD node, score
+						WHERE score >= $threshold
+						RETURN node, score
+						""", Map.of("indexName", INDEX_NAME, "numberOfNearestNeighbours", k, "embeddingValue",
+						embedding, "threshold", threshold))
+				.list(Neo4jVectorStore::recordToDocument);
+		}
+	}
+
+	@Override
+	public void afterPropertiesSet() {
```

Comment:
Consider using the `SmartLifecycle` instead of `InitializingBean`. 
Check the PgVectorStore impl. for a reference. Mind that the SmartLifecycle should support multiple start/stop cycles.

Reply:
I thought that the `SmartLifecycle` is not needed for our operations. The operations are idempotent.
The only reason I can come up with is based on the flag you are using in `PgVectorStore`: Remove everything related to the `Document` from the store. I thought it was only introduced for testing. If it is a valid production use-case, I am happy to introduce the lifecycle.

Reply:
Agree. I've been considering some future scenarios, where the lifecycle would help. 
But the InitializingBean does the job for now, we can think about the lifecycles in the future. Furthermore both could be complimentary.